dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/ExceptionHelper.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright 2024 OICR and UCSC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.dockstore.webservice.helpers;
+
+import jakarta.persistence.PersistenceException;
+import java.util.ArrayList;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Optional;
+import org.apache.http.HttpStatus;
+import org.hibernate.exception.ConstraintViolationException;
+
+/**
+ * Helper class that generates information about an exception/throwable.
+ * The implementation is single-serve - that is, each instance operates
+ * on a particular exception/throwable that is passed to the constructor.
+ */
+public class ExceptionHelper {
+
+    private static final Optional<Info> NONE = Optional.empty();
+    private final Throwable throwable;
+    private final List<Throwable> throwables;
+
+    public ExceptionHelper(Throwable throwable) {
+        this.throwable = throwable;
+        this.throwables = listThrowables(throwable);
+    }
+
+    // This is an improved implementation of ExceptionUtils.getThrowableList():
+    // https://github.com/apache/commons-lang/blob/0fde05172e853c3dff55d1841ad21c8cce363259/src/main/java/org/apache/commons/lang3/exception/ExceptionUtils.java#L517-L524
+    // The original has O(N^2) runtime on average, where N is the number of throwables, which could be used as an attack.
+    @SuppressWarnings("checkstyle:IllegalType")
+    private static List<Throwable> listThrowables(Throwable throwable) {
+        final LinkedHashSet<Throwable> list = new LinkedHashSet<>();
+        while (throwable != null && !list.contains(throwable)) {
+            list.add(throwable);
+            throwable = throwable.getCause();
+        }
+        return new ArrayList<>(list);
+    }
+
+    /**
+     * Calculates useful information about the exception/throwable.
+     */
+    public Info info() {
+        return handleJavaThrowable()
+            .or(() -> handleConstraintViolationException())
+            .or(() -> handlePersistenceException())
+            .or(() -> result(HttpStatus.SC_BAD_REQUEST))
```

Comment:
Shouldn't the fallback be an internal server error? We don't know what happened; it's some sort of unexpected exception -- while it could be a bad request, we don't know for sure and should probably look at it (we'd get notified via CloudWatch alerts).

Reply:
Can see it both ways, but your take seems reasonable, let's do it.  Also, we should probably handle `CustomWebApplicationException` by passing those messages/codes get as is.  Will make those mods.